hazelcast/src/test/java/com/hazelcast/map/ExpirationListenerTest.java

```diff
@@ -113,4 +150,55 @@ public void entryExpired(EntryEvent event) {
             expirationEventCount.countDown();
         }
     }
+
+    private static class LockingOverIMapExpirationListener implements EntryExpiredListener {
+
+        private final IMap map;
+        private final AtomicInteger eventCounter;
+
+        LockingOverIMapExpirationListener(AtomicInteger eventCounter, IMap map) {
+            this.map = map;
+            this.eventCounter = eventCounter;
+        }
+
+        @Override
+        public void entryExpired(EntryEvent event) {
+            eventCounter.incrementAndGet();
+
+            Object key = event.getKey();
+            map.lock(key);
+            try {
+                // mimic a slow execution
+                sleepMillis(10);
+            } finally {
+                map.unlock(key);
+            }
+        }
+    }
+
+    private static class LockingOverCpSubsystemExpirationListener implements EntryExpiredListener {
+
+        private final CPSubsystem cpSubsystem;
+        private final AtomicInteger eventCounter;
+
+        LockingOverCpSubsystemExpirationListener(AtomicInteger eventCounter, CPSubsystem cpSubsystem) {
+            this.cpSubsystem = cpSubsystem;
+            this.eventCounter = eventCounter;
+        }
+
+        @Override
+        public void entryExpired(EntryEvent event) {
+            eventCounter.incrementAndGet();
+
+            Object key = event.getKey();
+            FencedLock lock = cpSubsystem.getLock(key.toString());
+            lock.lock();
```

Comment:
Is it Ok to use the CP subsystem Lock in an unsafe mode?

Reply:
Yes, IMap lock has similar safety guarantees with unsafe locks of cp. It should be fine as a workaround for the main issue here.